File: src/node/spinner.js
/**
 * Terminal Spinner for Node.js
 * Developed by Ahmar Hussain (Stackaura)
 * https://www.stackaura.com/
 */

const SPINNERS = {
    dots: {
        interval: 80,
        frames: ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"]
    },
    line: {
        interval: 130,
        frames: ["-", "\\", "|", "/"]
    },
    classic: {
        interval: 200,
        frames: ["[    ]", "[=   ]", "[==  ]", "[=== ]", "[ ===]", "[  ==]", "[   =]", "[    ]", "[   =]", "[  ==]", "[ ===]", "[====]", "[=== ]", "[==  ]", "[=   ]"]
    }
};

class Spinner {
    constructor(type = 'dots') {
        this.spinner = SPINNERS[type] || SPINNERS.dots;
        this.frameIndex = 0;
        this.timer = null;
    }

    start(message = '') {
        process.stdout.write('\u001B[?25l'); // Hide cursor
        this.timer = setInterval(() => {
            const frame = this.spinner.frames[this.frameIndex];
            process.stdout.write(`\r${frame} ${message}`);
            this.frameIndex = (this.frameIndex + 1) % this.spinner.frames.length;
        }, this.spinner.interval);
    }

    stop(finalMessage = '') {
        clearInterval(this.timer);
        process.stdout.write(`\r${finalMessage}\n`);
        process.stdout.write('\u001B[?25h'); // Show cursor
    }
}

module.exports = Spinner;

File: src/python/spinner.py
import sys
import time
import threading

class Spinner:
    SPINNERS = {
        'dots': {
            'interval': 0.08,
            'frames': ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"]
        },
        'classic': {
            'interval': 0.2,
            'frames': ["[    ]", "[=   ]", "[==  ]", "[=== ]", "[ ===]", "[  ==]", "[   =]", "[    ]"]
        }
    }

    def __init__(self, type='dots'):
        self.spinner = self.SPINNERS.get(type, self.SPINNERS['dots'])
        self.stop_event = threading.Event()
        self.thread = None

    def _animate(self, message):
        while not self.stop_event.is_set():
            for frame in self.spinner['frames']:
                if self.stop_event.is_set():
                    break
                sys.stdout.write(f"\r{frame} {message}")
                sys.stdout.flush()
                time.sleep(self.spinner['interval'])

    def start(self, message=''):
        self.stop_event.clear()
        self.thread = threading.Thread(target=self._animate, args=(message,))
        self.thread.daemon = True
        self.thread.start()

    def stop(self, final_message=''):
        self.stop_event.set()
        if self.thread:
            self.thread.join()
        sys.stdout.write(f"\r{final_message}\n")
        sys.stdout.flush()
